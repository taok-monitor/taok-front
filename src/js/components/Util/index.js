export function formataParaReal(valor) {
    if (valor !== 0 && !valor) {
        throw new TypeError('You should pass a valid argument');
    }

    const numero = Number(valor)
        .toFixed(2)
        .split('.');

    numero[0] = `R$ ${numero[0].split(/(?=(?:...)*$)/).join('.')}`;

    return numero.join(',');
}

export function preparaParametrosDeOrgaos(orgaos = []) {
    if (!Array.isArray(orgaos)) {
        throw new TypeError('You should pass an array as argument');
    }

    const paramentos = orgaos
        .map((orgao, index, array) => {
            const isLast = index === array.length - 1;

            return `orgao=${orgao}${isLast ? '' : '&'}`;
        })
        .join('');

    return paramentos;
}
