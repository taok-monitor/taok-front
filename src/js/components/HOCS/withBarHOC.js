import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';

import * as Util from '../Util';

const legendas = [
    {
        label: 'FAS',
        legenda: 'FUNDO MUNICIPAL DE ASSISTENCIA SOCIAL',
    },
    {
        label: 'FMC',
        legenda: 'FUNDO MUNICIPAL DE CULTURA',
    },
    {
        label: 'FMDCA',
        legenda:
            'FUNDO MUNICIPAL DE DEFESA DOS DIREITOS DA CRIANCA E DO ADOLESCENTE',
    },
    {
        label: 'FMDSE',
        legenda: 'FUNDO MUNICIPAL DE DESENVOLVIMENTO ECONOMICO',
    },
    {
        label: 'FUNDURB',
        legenda: 'FUNDO MUNICIPAL DE DESENVOLVIMENTO URBANO',
    },
    {
        label: 'FMDD',
        legenda: 'FUNDO MUNICIPAL DE DIREITOS DIFUSOS',
    },
    {
        label: 'FME',
        legenda: 'FUNDO MUNICIPAL DE EDUCACAO',
    },
    {
        label: 'FMHIS',
        legenda: 'FUNDO MUNICIPAL DE HABITACAO DE INTERESSE SOCIAL',
    },
    {
        label: 'FEMS',
        legenda: 'FUNDO MUNICIPAL DE SAUDE',
    },
    {
        label: 'FMDPI',
        legenda: 'FUNDO MUNICIPAL DOS DIREITOS DA PESSOA IDOSA',
    },
    {
        label: 'FMDPD',
        legenda:
            'FUNDO MUNICIPAL PARA PROMOCAO DOS DIREITOS DAS PESSOAS COM DEFICIENCIA',
    },
    {
        label: 'GP',
        legenda: 'GABINETE DO PREFEITO',
    },
    {
        label: 'GVP',
        legenda: 'GABINETE DO VICE-PREFEITO',
    },
    {
        label: 'GMF',
        legenda: 'GUARDA MUNICIPAL DE FORTALEZA',
    },
    {
        label: 'HDMSJB',
        legenda: 'HOSPITAL DISTRITAL MARIA JOSE BARROSO DE OLIVEIRA',
    },
    {
        label: 'HEBO',
        legenda: 'HOSPITAL DISTRITAL EDMILSON BARROS DE OLIVEIRA',
    },
    {
        label: 'HDEAYRE',
        legenda: 'HOSPITAL DISTRITAL EVANDRO AYRES DE MOURA',
    },
    {
        label: 'HDGM/BC',
        legenda: 'HOSPITAL DISTRITAL GONZAGA MOTA/BARRA DO CEARA',
    },
    {
        label: 'HDGM/JW',
        legenda: 'HOSPITAL DISTRITAL GONZAGA MOTA/JOSE WALTER',
    },
    {
        label: 'HDGM/ME',
        legenda: 'HOSPITAL DISTRITAL GONZAGA MOTA/MESSEJANA',
    },
    {
        label: 'HDNSCON',
        legenda: 'HOSPITAL DISTRITAL NOSSA SENHORA DA CONCEICAO',
    },

    {
        label: 'HMZAN',
        legenda: 'HOSPITAL E MATERNIDADE DRA ZILDA ARNS NEUMANN',
    },
    {
        label: 'CROA',
        legenda: 'HOSPITAL LUCIA DE FATIMA/CROA',
    },
    {
        label: 'IJF',
        legenda: 'INSTITUTO DR. JOSE FROTA',
    },
];

function calculaAcumulado(valores = []) {
    const total = valores.reduce(
        (valorFinal, atual) => valorFinal + atual,
        0.0
    );

    return Util.formataParaReal(total);
}

function encontraLegenda(label) {
    const legendaEncontrada = legendas.find(legenda => legenda.label === label);

    return legendaEncontrada ? legendaEncontrada.legenda : '';
}

export function getFormattedLabel(tooltipItem, data) {
    const { data: dataValues } = data.datasets[tooltipItem.datasetIndex];
    const label = encontraLegenda(data.labels[tooltipItem.index]);
    const value = dataValues[tooltipItem.index];

    return `${label}: ${Util.formataParaReal(value)}`;
}

export function withBarHOC(Component) {
    class BarHOC extends React.Component {
        constructor(props) {
            super(props);

            this.state = {
                labels: [],
                valores: [],
                orgaos: props.orgaos,
                mesAnoInicial: props.mesAnoInicial,
                mesAnoFinal: props.mesAnoFinal,
                totalAcumulado: 0,
            };
        }

        componentDidMount() {
            this.obtemDados();
        }

        componentWillReceiveProps(nextProps) {
            const { mesAnoInicial, mesAnoFinal, orgaos } = nextProps;

            this.setState(
                {
                    mesAnoInicial,
                    mesAnoFinal,
                    orgaos,
                },
                this.obtemDados
            );
        }

        obtemDados() {
            const { api } = this.props;
            const { mesAnoInicial, mesAnoFinal, orgaos } = this.state;
            const parametrosDeOrgaosFormatados = Util.preparaParametrosDeOrgaos(
                orgaos
            );

            axios
                .get(
                    `${api}${mesAnoInicial}/${mesAnoFinal}?${parametrosDeOrgaosFormatados}`
                )
                .then(res => {
                    this.setState({
                        labels: res.data[0].labels,
                        valores: res.data[0].valores,
                        totalAcumulado: calculaAcumulado(res.data[0].valores),
                    });
                });
        }

        render() {
            return (
                <Component
                    {...this.props}
                    {...this.state}
                    getFormattedLabel={getFormattedLabel}
                />
            );
        }
    }

    BarHOC.propTypes = {
        orgaos: PropTypes.arrayOf(PropTypes.any).isRequired,
        mesAnoInicial: PropTypes.string.isRequired,
        mesAnoFinal: PropTypes.string.isRequired,
        api: PropTypes.string.isRequired,
    };

    return BarHOC;
}

export default withBarHOC;
