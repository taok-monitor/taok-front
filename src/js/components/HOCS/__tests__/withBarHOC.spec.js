/* eslint-disable react/prop-types */
import React from 'react';
import { render, wait, cleanup, fireEvent } from '@testing-library/react';
import { withBarHOC, getFormattedLabel } from '../withBarHOC';
import * as Util from '../../Util';

jest.mock('axios', () => ({
    get: () => {
        return Promise.resolve({
            data: [
                {
                    labels: ['HMZAN'],
                    valores: [46445.82],
                },
            ],
        });
    },
}));

const defaultProps = {
    orgaos: [],
    mesAnoInicial: '062019',
    mesAnoFinal: '072019',
    api: 'somewhere.com',
};

const tooltipItem = {
    index: 0,
    datasetIndex: 0,
};

const data = {
    labels: ['IJF'],
    datasets: [
        {
            data: [1000],
        },
    ],
};

/**
 * This is a fake component for us simulate the behaviour from
 * `withBarHOC`.
 */
const MockedBar = withBarHOC(
    class extends React.Component {
        constructor() {
            super();

            this.state = {};
            this.handleOnMouseEnter = this.handleOnMouseEnter.bind(this);
        }

        handleOnMouseEnter() {
            this.setState({
                formattedLabel: this.props.getFormattedLabel(tooltipItem, data),
            });
        }

        render() {
            return (
                <div>
                    <p>Total acumulado: {this.props.totalAcumulado}</p>
                    <p>Mes e ano inicial: {this.props.mesAnoInicial}</p>
                    <p>Mes e ano final: {this.props.mesAnoFinal}</p>
                    <button
                        type="button"
                        onMouseEnter={this.handleOnMouseEnter}
                    >
                        onMouseEnter to call getFormattedLabel
                    </button>
                    {this.state.formattedLabel && (
                        <p>{this.state.formattedLabel}</p>
                    )}
                </div>
            );
        }
    }
);

afterEach(cleanup);

describe('withBarHOC', () => {
    it('should render correctly and pass props down', async () => {
        const { container, getByText } = render(
            <MockedBar {...defaultProps} />
        );

        await wait(() => {
            expect(
                getByText('Total acumulado: R$ 46.445,82')
            ).toBeInTheDocument();
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    it('should get data again when props change', async () => {
        const { rerender, container, getByText } = render(
            <MockedBar {...defaultProps} />
        );

        await wait(() => {
            expect(
                getByText('Total acumulado: R$ 46.445,82')
            ).toBeInTheDocument();
            expect(getByText('Mes e ano inicial: 062019')).toBeInTheDocument();
            expect(getByText('Mes e ano final: 072019')).toBeInTheDocument();
            expect(container.firstChild).toMatchSnapshot();
        });

        rerender(
            <MockedBar
                {...defaultProps}
                mesAnoInicial="072019"
                mesAnoFinal="082019"
            />
        );

        await wait(() => {
            expect(getByText('Mes e ano inicial: 072019')).toBeInTheDocument();
            expect(getByText('Mes e ano final: 082019')).toBeInTheDocument();
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    it('should call `getFormattedLabel` onMouseEnter', async () => {
        const { container, getByText } = render(
            <MockedBar {...defaultProps} />
        );

        const button = getByText('onMouseEnter to call getFormattedLabel');

        fireEvent.mouseEnter(button);

        await wait(() => {
            expect(
                getByText(getFormattedLabel(tooltipItem, data))
            ).toBeInTheDocument();
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});
