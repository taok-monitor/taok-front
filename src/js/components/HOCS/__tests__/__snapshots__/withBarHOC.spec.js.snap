// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withBarHOC should call \`getFormattedLabel\` onMouseEnter 1`] = `
<div>
  <p>
    Total acumulado: 
    R$ 46.445,82
  </p>
  <p>
    Mes e ano inicial: 
    062019
  </p>
  <p>
    Mes e ano final: 
    072019
  </p>
  <button
    type="button"
  >
    onMouseEnter to call getFormattedLabel
  </button>
  <p>
    INSTITUTO DR. JOSE FROTA: R$ 1.000,00
  </p>
</div>
`;

exports[`withBarHOC should get data again when props change 1`] = `
<div>
  <p>
    Total acumulado: 
    R$ 46.445,82
  </p>
  <p>
    Mes e ano inicial: 
    062019
  </p>
  <p>
    Mes e ano final: 
    072019
  </p>
  <button
    type="button"
  >
    onMouseEnter to call getFormattedLabel
  </button>
</div>
`;

exports[`withBarHOC should get data again when props change 2`] = `
<div>
  <p>
    Total acumulado: 
    R$ 46.445,82
  </p>
  <p>
    Mes e ano inicial: 
    072019
  </p>
  <p>
    Mes e ano final: 
    082019
  </p>
  <button
    type="button"
  >
    onMouseEnter to call getFormattedLabel
  </button>
</div>
`;

exports[`withBarHOC should render correctly and pass props down 1`] = `
<div>
  <p>
    Total acumulado: 
    R$ 46.445,82
  </p>
  <p>
    Mes e ano inicial: 
    062019
  </p>
  <p>
    Mes e ano final: 
    072019
  </p>
  <button
    type="button"
  >
    onMouseEnter to call getFormattedLabel
  </button>
</div>
`;
